src/clients/BundleDataClient/BundleDataClient.ts

```diff
@@ -1043,37 +1002,127 @@ export class BundleDataClient {
                 this.getRelayHashFromEvent(matchedDeposit) === relayDataHash,
                 "Deposit relay hashes should match."
               );
-
-              // slow fill requests for deposits from or to lite chains are considered invalid
-              if (matchedDeposit.fromLiteChain || matchedDeposit.toLiteChain) {
-                return;
-              }
-
               v3RelayHashes[relayDataHash].deposit = matchedDeposit;
 
-              // Note: we don't need to query for a historical fill at this point because a fill
-              // cannot precede a slow fill request and if the fill came after the slow fill request,
-              // we would have seen it already because we would have processed it in the loop above.
               if (
-                // Input and Output tokens must be equivalent on the deposit for this to be slow filled.
-                !this.clients.hubPoolClient.areTokensEquivalent(
-                  matchedDeposit.inputToken,
-                  matchedDeposit.originChainId,
-                  matchedDeposit.outputToken,
-                  matchedDeposit.destinationChainId,
-                  matchedDeposit.quoteBlockNumber
-                ) ||
+                !_canCreateSlowFillLeaf(matchedDeposit) ||
                 // Deposit must not have expired in this bundle.
-                slowFillRequest.fillDeadline < bundleBlockTimestamps[destinationChainId][1]
+                _depositIsExpired(matchedDeposit)
               ) {
-                // TODO: Invalid slow fill request. Maybe worth logging.
                 return;
               }
               validatedBundleSlowFills.push(matchedDeposit);
             }
           }
         );
 
+        // Deposits can be submitted an arbitrary amount of time after matching fills and slow fill requests.
+        // Therefore, let's go through each deposit in this bundle again and check a few things in order:
+        // - Has the deposit been filled ? If so, then we need to issue a relayer refund for
+        //   this "pre-fill" if the fill took place in a previous bundle.
+        // - Or, has the deposit expired in this bundle? If so, then we need to issue an expiry refund.
+        // - And finally, has the deposit been slow filled? If so, then we need to issue a slow fill leaf
+        //   for this "pre-slow-fill-request" if this request took place in a previous bundle.
+        const originBlockRange = getBlockRangeForChain(blockRangesForChains, originChainId, chainIds);
+
+        await mapAsync(
+          bundleDepositHashes.filter((depositHash) => {
+            const { deposit } = v3RelayHashes[depositHash];
+            return (
+              deposit &&
+              deposit.originChainId === originChainId &&
+              deposit.destinationChainId === destinationChainId &&
+              deposit.blockNumber >= originBlockRange[0] &&
```

Comment:
These last three checks about the block ranges and zero value deposits are redundant since we check them when adding to `bundleDepositHashes`. Do we want to keep them for the sake of being extra safe (e.g. if we change how we add stuff to `bundleDepositHashes` in the future)?

Reply:
good call will remove